Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/Solver.cs

```diff
@@ -210,12 +210,18 @@ protected virtual void Awake()
         }
 
         /// <summary>
-        /// Typically when a solver becomes enabled, it should update its internal state to the system, in case it was disabled far away
+        /// When a solver becomes enabled, it must make sure it is registered at its <see cref="SolverHandler"/>.
+        /// Also it should update its internal state to the system, in case it was disabled far away.
         /// </summary>
         protected virtual void OnEnable()
         {
             if (SolverHandler != null)
             {
+                if(!SolverHandler.Solvers.Contains(this))
+                { 
+                    SolverHandler.Solvers = SolverHandler.GetComponents<Solver>();
+                }
```

Comment:
A couple of thoughts here - instead of having consumers of an object determine what solvers it manages, I think it would be better to instead of a system where instead the Solver will tell the SolverHandler that "hey, you should update your list" and the solver handler, in its LateUpdate() call, can then do the work to refresh its list once.

This has some benefits of only having to do the GetComponents<> call once if enabling a bunch of solvers in the same update loop, and also makes it so that you avoid propagating bugs in the Solvers themselves up the chain. What I mean by this is, if there were a bug in some solver code that updated SolverHandler.Solvers to be empty, it would now break all other solvers attached to that handler.

If instead, the was just an api that was like:

SolverHandler.QueueSolverListUpdate();

Then the surface area for error would also be less.

I *think* that this would also eliminate the need to do the:

public IReadOnlyCollection<Solver> Solvers

Change in SolverHandler.cs

So suggestion here is to:

protected virtual void OnEnable()
        {
            if (SolverHandler != null)
            {
                if(!SolverHandler.Solvers.Contains(this))
                { 
                    SolverHandler.QueueSolverListUpdate();
                }

And then in solver handler, in LateUpdate, just have the code that's currently in this section there instead.

Reply:
I would actually go a different route.

The GetComponents<Solver>() is very dangerous. and as @wiwei said bad performance. Furthermore, the SolverManager is supposed to manage the solvers...but a Solver telling a manager exactly what to manage seems very weird and problematic. It's probably better for a Solver to just say hey! manage me in a sort of auto-registeration

1) So SolverHandler would be updated to not set a default Solver list 
2) In Solver.Start(), have the solver register itself with the SolverHandler
3) In OnDestroy, solver unregisters itself

We should keep the Solver property as readonly. There is the SetSolvers() function if a higher order manager system wants to override the list

```csharp
public class solver
{
	Start()
	{
		if (SolverHandler != null)
		{
			SolverHandler.RegisterSolver(this)
		}
	}

	OnDestroy()
	{
		if (SolverHandler != null)
		{
			SolverHandler.UnregisterSolver(this)
		}
	}
}

public class SovlerHandler
{
	RegisterSolver(Solver s)
	{
		if s != null & !solvers.Contains(s)
		{
			solvers.Add(s);
		}
	}

	UnregisterSolver(Solver s)
	{
		if (s != null && solvers.Contains(s))
		{
			solvers.Remove(s);
		}
	}
}

```

Reply:
That was my thought as well, but the fact that `SolverHandler.Solvers` is not readonly, but others can programatically change the list, and the Docs state that the update order shall be the order of components in the Inspector confused me. I will change this as it is indeed less error prone.

Reply:
Yes the docs will need to be updated as well. 

The GetComponents call by Unity will return components in the order that they appear in the inspector. SolverHandler by default calls GetComponents < Solver >